src/charm.py

```diff
@@ -121,92 +123,107 @@ def _get_database_relation_info(self) -> dict:
-            logger.info("Missing database details. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting for database creation")
-            return
-
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        self._update_layer()
+        self.unit.status = MaintenanceStatus("Configuring resources")
+
+        self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
+        logger.info("Pebble plan updated with new configuration, replanning")
+        try:
+            self._container.replan()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+        if self.database.is_database_created():
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            self._container.start(self._container_name)
+            self.unit.status = ActiveStatus()
+
+    def _on_database_created(self, event) -> None:
+        """Event Handler for database created event."""
+        logger.info("Retrieved database details")
 
         if not self.unit.is_leader():
             # TODO: Observe leader_elected event
             logger.info("Unit does not have leadership")
             self.unit.status = WaitingStatus("Waiting for leadership")
             return
 
-        self._run_sql_migration()
-
-        self.unit.status = ActiveStatus()
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Hydra container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
+            return
 
-    def _run_sql_migration(self) -> None:
-        """Runs a command to create SQL schemas and apply migration plans."""
-        process = self._container.exec(
-            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
-            timeout=20.0,
+        self.unit.status = MaintenanceStatus(
+            "Configuring container and resources for database connection"
         )
+
         try:
-            stdout, _ = process.wait_output()
-            logger.info(f"Executing automigration: {stdout}")
-        except ExecError as err:
-            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
-            self.unit.status = BlockedStatus("Database migration job failed")
+            self._container.get_service(self._container_name)
+            logger.info("Updating Hydra config and restarting service")
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            self._run_sql_migration()
+            self._container.restart(self._container_name)
```

Comment:
We could push the pebble layer again here and update it to `"startup": "enabled"`